modules/bim/app/controllers/bim/bcf/api/v2_1/comments/api.rb

```diff
@@ -0,0 +1,93 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2021 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Bim::Bcf::API::V2_1
+  module Comments
+    class API < ::API::OpenProjectAPI
+      resources :comments do
+        helpers do
+          def all_comments
+            @issue.comments.includes(:journal, :issue, :viewpoint)
+          end
+
+          def get_viewpoint_by_uuid(viewpoint_uuid)
+            if viewpoint_uuid == nil
+              nil
+            else
+              @issue.viewpoints.find_by(uuid: viewpoint_uuid) || ::Bim::Bcf::NonExistentViewpoint.new
```

Comment:
I'm a bit torn with this approach. I can see why you did it: You get validations in the contract and get all errors returned to the user whereas failing the finder would fail the API request with a single error instead.

But you have to instantiate a new record subclass that feels a bit unnecessary. I don't have a better idea, however.

Reply:
I could fail here but then I would kind of circumvent the role of the contract for this route. This is mostly a design decision. From functionality I (and the API by specification) could live with both.